File: NOIP/NOIP复赛/NOIP2010/提高组参考程序/Pascal/translate.pas
program translate(input,output);
  const
    maxn=1000;
  var
    n,m,f,r,x,i,ans:longint;
    q,a:array[1..maxn] of longint;

  procedure push(x:longint);
    begin
      if a[x]=1 then exit;
      inc(r);q[r]:=x;a[x]:=1;inc(ans);
      if r-f+1>m then begin
        a[q[f]]:=0;
        inc(f);
      end;
    end;

  begin
    assign(input,'translate.in');reset(input);
    assign(output,'translate.out');rewrite(output);
    fillchar(a,sizeof(a),0);
    readln(m,n);
    ans:=0;
    for i:=1 to n do begin
      read(x);
      push(x);
    end;
    writeln(ans);
    close(input);close(output);
  end.

File: NOIP/NOIP复赛/NOIP2010/提高组参考程序/C++/translate.cpp
#include <iostream>
#include <fstream>

const int maxn=1000;
int f,r,n,m,ans;
int q[maxn+1],a[maxn+1];using namespace std;

void push(int x)
{
	if(a[x]!=0) return;
	r++;q[r]=x;a[x]=1;
	ans++;
	if(r-f+1>m){
		a[q[f]]=0;
		f++;
	}
}
int main()
{
	ifstream infile("translate.in",ios::in);
	ofstream outfile("translate.out",ios::out);
	memset(a,0,sizeof(a));
	int x;
	infile>>m>>n;
	f=1;r=0;ans=0;
	for(int i=1;i<=n;i++){
		infile>>x;
		push(x);
	}
	outfile<<ans<<endl;
	infile.close();
	outfile.close();
	return 0;}
